sdk/src/main/java/io/opentelemetry/sdk/trace/EventSdk.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.trace;
+
+import com.google.auto.value.AutoValue;
+import io.opentelemetry.trace.AttributeValue;
+import io.opentelemetry.trace.Event;
+import java.util.Collections;
+import java.util.Map;
+import javax.annotation.concurrent.Immutable;
+
+/** Implementation of {@link Event}. */
+@Immutable
+@AutoValue
+public abstract class EventSdk implements Event {
```

Comment:
Some questions:
1. Why is this public?
1. Why do we not have a TimedEvent class that includes the nanotime as well + oneof(string + attributes, or Event), because we need to store the time when the event was recorded as well.

Reply:
> Why is this public?

You're right this doesn't need to be public, updated.

> Why do we not have a TimedEvent class that includes the nanotime as well + oneof(string + attributes, or Event), because we need to store the time when the event was recorded as well.

We already have a concrete [TimeEvent](https://github.com/open-telemetry/opentelemetry-java/blob/792ffa9c5965796f940631fe8b0abd2a28c0fce9/api/src/main/java/io/opentelemetry/trace/SpanData.java#L225) class in the API, but we don't have an implementation for `Event`. The use case is 
```java
  @Override
  public void addEvent(String name) {
    addEvent(EventSdk.create(name));
  }

  @Override
  public void addEvent(String name, Map<String, AttributeValue> attributes) {
    addEvent(EventSdk.create(name, attributes));
  }

  @Override
  public void addEvent(Event event) {
    ...
    TimedEvent.create(timestamp, event);
  }
```

Reply:
We also have event there https://github.com/open-telemetry/opentelemetry-java/blob/792ffa9c5965796f940631fe8b0abd2a28c0fce9/api/src/main/java/io/opentelemetry/trace/SpanData.java#L185

But I think we should not use that because it forces 2 allocations one for event one for the TimeEvent. Also the TimeEvent uses Timestamp not nanotime which forces one more allocation. I was thinking to have something like

```java
abstract static class TimedEvent {
  abstract long getNanotime();

  abstract String getEventName();

  abstract Map<String, AttributeValue> getEventAttributes();

  create(long nanotime, String name, Map<String, AttributeValue> attributes) {
    return new AutoValue_TimedEvent_RawTimeEvent(nanotime, name, attributes);
  }

  create(long nanotime, Event event) {
    return new AutoValue_TimedEvent_TimeEventWithEvent(nanotime, event);
  }

  @AutoValue
  static RawTimedEvent extends TimedEvent {
  }

  @AutoValue
  static TimeEventWithEvent extends TimedEvent {
    String getEventName() {
      getEvent().getName();
    }

    Map<String, AttributeValue> getEventAttributes() {
      getEvent().getName();
    }

    abstract Event getEvent();
  }
}
```

Reply:
Good points with object allocation, updated to use `TimedEvent` instead in https://github.com/open-telemetry/opentelemetry-java/pull/354/commits/d04717f8556521605a9598e874596cb79255f202. PTAL